File: algorithm/207. Course Schedule/cpp/CourseSchedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        int n = prerequisites.size();
        vector<vector<int>> graph(numCourses);
        // build graph
        for(int i = 0; i < n; i++){
            graph[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }
        vector<bool> visited(numCourses); // used in each dfs to check if there is a cycle
        vector<bool> done(numCourses);  // record the courses that has been checked
        for(int i = 0; i < numCourses; i++){
            if(done[i]) continue;
            if(!dfs(graph,visited,done,i)) return false;
            done[i] = true;
        }
        return true;
    }
private:
    bool dfs(vector<vector<int>>& graph, vector<bool>& visited, vector<bool>& done, int course){
        if(visited[course]) return false;
        visited[course] = true;
        for(int next:graph[course]){
            if(!done[next]&&!dfs(graph,visited,done,next)) return false;
            done[next] = true;
        }
        visited[course] = false;
        return true;
    }
};

File: algorithm/207. Course Schedule/java/CourseSchedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // convert to adjacency list
        List<Integer>[] graph = new ArrayList[numCourses];
        // Initialization
        for(int i = 0; i < numCourses; i++) graph[i] = new ArrayList();
        // Build Graph
        for(int i = 0; i < prerequisites.length; i++){
            graph[prerequisites[i][0]].add(prerequisites[i][1]);
        }
        
        boolean[] visited = new boolean[numCourses];
        boolean[] done = new boolean[numCourses];
        for(int i = 0; i < numCourses; i++){
            if(done[i]) continue;
            if(!recur(graph, visited, done, i)) return false;
            done[i] = true;
        }
        return true;
    }
    
    private boolean recur(List<Integer>[] graph, boolean[] visited, boolean[] done, int curCourse){
        if(visited[curCourse]) return false;
        visited[curCourse] = true;
        for(int course:graph[curCourse]){
            if(!done[course] && !recur(graph,visited,done,course)){
                return false;
            }
            done[course] = true;
        }
        visited[curCourse] = false;
        return true;
    }
}
